[ao][refactor] add is_dynamic flag to determine quantization type

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75263

Summary: The is_dynamic flag was added to observers, the relevant
qconfigs and inherited from observers by fake_quants so that it can be
used to determine the quantization type (static vs dyamic) for the
observed module. compute_dtype was also removed since that is how quant
type was previously determined.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35406334](https://our.internmc.facebook.com/intern/diff/D35406334)

HDCharles committed:
[ao][refactor] add is_dynamic flag to determine quantization type

Summary: The is_dynamic flag was added to observers, the relevant
qconfigs and inherited from observers by fake_quants so that it can be
used to determine the quantization type (static vs dyamic) for the
observed module.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35406334).

jerryzh168 reviewed (approved):
lg thanks, please fix lint before landing

HDCharles committed:
Update on "[ao][refactor] add is_dynamic flag to determine quantization type"


Summary: The is_dynamic flag was added to observers, the relevant
qconfigs and inherited from observers by fake_quants so that it can be
used to determine the quantization type (static vs dyamic) for the
observed module. with the addition of is_dynamic, compute_dtype needs to be refactored out since it does the same job.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35406334](https://our.internmc.facebook.com/intern/diff/D35406334)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fake_quantize.py:
```diff
@@ -347,7 +348,7 @@ def forward(self, X: torch.Tensor) -> torch.Tensor:
 
 default_dynamic_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, quant_min=0, quant_max=255,
-                                                    dtype=torch.quint8, averaging_constant=1)
+                                                    dtype=torch.quint8, averaging_constant=1, is_dynamic=True)
```
if we inherit this from observer I think we can remove this flag from fake quantize api

jerryzh168 replied:
should be configured like this:

FakeQuantize.with_args(observer=MovingAverageMinMaxObserver.with_args(is_dynamic=True), ...)

HDCharles replied:
I think all arguments aside from quant_min and quant_max are passed to the observer via observer kwargs here: https://www.internalfb.com/code/fbsource/[205ceaf0d5c23e4fdc62814a3622c745703e6b58]/fbcode/caffe2/torch/ao/quantization/fake_quantize.py?lines=140

since things like averaging_constant are passed via that method I figured this would be fine. It seems weird to me to pass 1 via that method and another by directly modifying the observer argument.

jerryzh168 replied:
I see, maybe we can postpone this, but I feel the final state should be we just modify the config for observer directly

HDCharles committed:
Update on "[ao][refactor] add is_dynamic flag to determine quantization type"


Summary: The is_dynamic flag was added to observers, the relevant
qconfigs and inherited from observers by fake_quants so that it can be
used to determine the quantization type (static vs dyamic) for the
observed module. compute_dtype was also removed since that is how quant
type was previously determined.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35406334](https://our.internmc.facebook.com/intern/diff/D35406334)

[ghstack-poisoned]

vkuzo reviewed (changes requested):
I didn't review the code, but asking for changes to add more context on the reasoning for this behind the PR summary.  Can we please document what the current state of qconfig is, and how this change impacts the design, before going forward?

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -349,6 +351,7 @@ class MinMaxObserver(_ObserverBase):
         quant_min: Minimum quantization value. If unspecified, it will follow the 8-bit setup.
         quant_max: Maximum quantization value. If unspecified, it will follow the 8-bit setup.
         eps: Epsilon value for float32, Defaults to `torch.finfo(torch.float32).eps`.
+        is_dynamic: Indicator that dynamic quantization will be used for observed module.
```
maybe we can give a more detailed docstring for this flag, e.g. specifically mentioning how this flag impacts the conversion from observer to quantize op.
(1). when is_dynamic is False, the observer will be converted to quantize_per_tensor - dequantize ops in convert
(2). when is_dynamic is True, the observer will be converted to quantize_per_tensor_dynamic - dequantize ops

jerryzh168 commented:
@HDCharles let's rename this flag to "convert_to_quantize_per_tensor_dynamic" to make it more explicit

also we can add this example comparison between static and dynamic quantization to PR summary:

Static Quantization
fp32 model:
Input → conv → output

Observed
Input → observer.with_args(is_dynamic=False) →  conv → output →  observer.with_args(is_dynamic=False) → …

Reference:
Input → quantize_per_tensor → dequantize → conv → output → quantize_per_tensor → dequantize → …

Dynamic Quantization

fp32 model:
Input → conv → output

Observed
Input → observer.with_args(is_dynamic=True) →  conv → output →  observer.with_args(is_dynamic=True) → …

Reference:
Input → quantize_per_tensor_dynamic → dequantize → conv → output → quantize_per_tensor_dynamic → dequantize → …

HDCharles commented:
> I didn't review the code, but asking for changes to add more context on the reasoning for this behind the PR summary. Can we please document what the current state of qconfig is, and how this change impacts the design, before going forward?

Yeah, will do

jerryzh168 commented:
discussed in the doc (https://docs.google.com/document/d/1_Snruvj8THv9KMU_AE6KkdAX8tUGfNJGDxSN0biqXoY/edit#), maybe we can use: `replacement_quant_is_dynamic` instead to keep the name shorter

HDCharles committed:
Update on "[ao][refactor] add is_dynamic flag to determine quantization type"


Summary: The is_dynamic flag was added to observers, the relevant
qconfigs and inherited from observers by fake_quants so that it can be
used to determine the quantization type (static vs dyamic) for the
observed module. compute_dtype was also removed since that is how quant
type was previously determined.

Test Plan: python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35406334](https://our.internmc.facebook.com/intern/diff/D35406334)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -98,10 +98,10 @@ class ObserverBase(ABC, nn.Module):
     """
 
-    def __init__(self, dtype, is_dynamic=None):
+    def __init__(self, dtype, replacement_quant_is_dynamic=None):
```
should we just set default to True, just to reduce the possible values of `replacement_quant_is_dynamic`

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -514,14 +493,8 @@ def maybe_insert_input_observer_for_arg_or_kwarg(
-            # of dynamic quantization, e.g. dynamically quantizing second input, third
-            # input etc.
-            (arg_as_input_target_compute_dtype in [torch.quint8, torch.int8, torch.float16]) and arg is node.args[0]
+            qconfig.activation().replacement_quant_is_dynamic and arg is node.args[0]
```
is `arg is node.args[0]` needed?

jerryzh168 replied:
oh I think it's fine to keep this, since the qconfig is assuming this right now

jerryzh168 replied:
maybe we can return `replacement_quant_is_dynamic` from `get_qconfig_dtypes` and store this information together with other dtype information, in L358, so that we don't need to query qconfig again here.

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -722,7 +695,7 @@ def maybe_insert_output_observer_for_node(
     # if needed, they are inserted inside the standalone module
     should_insert_observer = should_insert_observer and \
-        (not is_standalone_module)
+        (not is_standalone_module) and not activation_is_dynamically_quantized(qconfig)
```
why we don't insert observer for dynamically quantized tensors?

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -223,35 +221,34 @@ def op_is_int8_dynamically_quantized(qconfig) -> bool:
+        activation_is_dynamically_quantized(qconfig)
     )
 
 def get_qconfig_dtypes(qconfig):
```
should we return is_replacement_quant_dynamic as well

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -350,14 +350,13 @@ def get_target_activation_dtype_for_node(
         if qconfig is not None:
             if qhandler is not None and qhandler.input_output_observed():
-                act_dtype, weight_dtype, act_compute_dtype = \
+                act_dtype, weight_dtype = \
```
should we indicate it's dynamic quantization here?

HDCharles commented:
I'm running into the following error in DBR:

when I run `python test/test_quantization.py TestQuantizeDBRIndividualOps.test_linear_dynamic`

when run on master the input_tensor_infos and output_tensor_infos are all float32 whereas here it changes to torch.quint8 which makes it pass the 
`if input_arg.inf_dtype != output_dtype:` in the same function. Normally i'd replace that with a check for is_dynamic but we also don't seem to have access to that information at that point and I'm not sure thats the intent of the flow here.
I'm unsure as to the best way to deal with this, 

```
True True False
are we here 1 (type): <class 'torch.nn.modules.linear.Linear'>
     (fqn): linear
     (input_tensor_infos): [QTensorInfo(id=0, orig_dtype=torch.float32, inf_dtype=torch.float32)]
     (output_tensor_infos): [QTensorInfo(id=1, orig_dtype=torch.float32, inf_dtype=torch.quint8)]
are we here 2
are we here 3 torch.float32 torch.quint8
are we here 4
F
======================================================================
FAIL: test_linear_dynamic (quantization.dbr.test_quantize_dbr.TestQuantizeDBRIndividualOps)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/cdhernandez/pytorch/test/quantization/dbr/test_quantize_dbr.py", line 195, in test_linear_dynamic
    self._test_auto_tracing(m, qconfig, (torch.randn(1, 1, 1, 1),))
  File "/data/users/cdhernandez/pytorch/test/quantization/dbr/test_quantize_dbr.py", line 77, in _test_auto_tracing
    mq = _quantize_dbr.convert(mp)
  File "/data/users/cdhernandez/pytorch/torch/ao/quantization/_quantize_dbr.py", line 142, in convert
    model = add_auto_convert(model)
  File "/data/users/cdhernandez/pytorch/torch/ao/quantization/_dbr/auto_trace.py", line 699, in add_auto_convert
    attach_op_convert_info_to_model(module)
  File "/data/users/cdhernandez/pytorch/torch/ao/quantization/_dbr/model_utils.py", line 145, in attach_op_convert_info_to_model
    qstate.calculate_op_convert_info(seen_q_op_info)
  File "/data/users/cdhernandez/pytorch/torch/ao/quantization/_dbr/quantization_state.py", line 577, in calculate_op_convert_info
    get_input_args_quant_dequant_info(
  File "/data/users/cdhernandez/pytorch/torch/ao/quantization/_dbr/utils.py", line 700, in get_input_args_quant_dequant_info
    assert tensor_id in tensor_id_to_scale_zp
AssertionError
```

HDCharles commented:
@vkuzo yeah that behavior has changed, we removed compute dtype and returned the behavior of dtype to how it was originally so dynamci quantization has activation observer with dtype q(u)int8 and the is_dynamic flag set.

vkuzo commented:
> @vkuzo yeah that behavior has changed, we removed compute dtype and returned the behavior of dtype to how it was originally so dynamci quantization has activation observer with dtype q(u)int8 and the is_dynamic flag set.

lol, maybe I didn't understand what it did originally then.  To me it makes sense for `dtype` to mean "this is going to be the dtype of the tensor which will flow through this part of the model during inference".  It's surprising that it would not be torch.float for the output of qlinear_dynamic.

What specifically is "observer.dtype" after this PR?  Can we explain it in a paragraph?

vkuzo commented on torch/ao/quantization/utils.py:
```diff
@@ -223,35 +221,34 @@ def op_is_int8_dynamically_quantized(qconfig) -> bool:
         get_qconfig_dtypes(qconfig)
     return (
-        activation_dtype is torch.float and
+        activation_is_int8_quantized(qconfig) and
```
this is a bit confusing.  `activation` here is talking about the activation after quantization.  It's not talking about the reference model - that's an intermediate step which is not exposed to most users.
